src/python/pants/notes/2.16.x.md

```diff
@@ -6,18 +6,75 @@
 
 The new `env` function in BUILD files allows access to environment variables in BUILD files.
 
+Run `pants help symbols` for documentation of all builtin symbols and macros provided by prelude files
+which are usable in BUILD files.
+
+The visibility rules syntax now supports matching on target names. The rules glob syntax also now uses
+the same syntax as target selectors.
+
 ### Backends
 
 #### Python
 
-The Python backend has two new linters:
+The Python backend added or improved support for various tools including:
 
-[`pydocstyle`](https://www.pydocstyle.org/) is now supported for linting Python doc strings. Enable the
+- [`pydocstyle`](https://www.pydocstyle.org/) is now supported for linting Python doc strings. Enable the
 `pants.backend.python.lint.pydocstyle` backend to add this support.
 
-[`ruff`](https://github.com/charliermarsh/ruff) is now supported for linting and formatting Python code. Enable the
+- [`ruff`](https://github.com/charliermarsh/ruff) is now supported for linting and formatting Python code. Enable the
 `pants.backend.experimental.python.lint.ruff` backend to add this support.
 
+- Pants now understands how to parse dependencies of projects using the OpenStack/Stevedore ((https://github.com/openstack/stevedore)
+project. Enable the `pants.backend.python.framework.stevedore` backend to add this support.
+
+- More Python target types now have the `environment` field and thus can execute in non-local environments.
+
+Deprecations:
+
+- The `[python].interpreter_constraints` option is now deprecated and will be removed in Pants v2.17.x. 
+This option configured default Python interpreter constraints for Python-related targets which did not
+otherwise specify their own interpreter constraints. Instead, use the `interpreter_constraints` field available on
+various Python-relate target types to explicitly specify what Python interpreters your code is compatible with.  
+
+#### Shell & Adhoc
```

Comment:
@chrisjrn / @benjyw: Is there an example repo for the new/improved Shell/Adhoc changes? It would be good to link it in this section.

Also can you suggest what kinds of workflows are amenable to this support?

Reply:
I took a stab at answering the second question in https://github.com/pantsbuild/pants/pull/18422.